File: 3_Signalanalyse/Python/7_WVD/wvd.py
import numpy as np
import math as math


def wvd(signal_in, resolution, win_size, sampl_freq):
    """calculate the Wigner-Ville distribution (WVD) of signal_in

    Args:
        signal_in (array_like): signal from that the cwt should be calculated
        resolution (float): resolution, number of samples between windows (for full resolution:  1)
        winSize (flaot): window size, becomes length of frequency axis (a good default: len(signalIn)/2)
        sampl_frequ (float): samples per second of signal
    Returns:
        cwt(array_like): WVD of signal_in
        ff(array_like): frequency array
        tt(array_like): time array
    """
    number_pts = math.floor(math.floor(len(signal_in)/resolution)/2)*2

    win_size = math.floor(win_size)

    odd_win = (math.floor((win_size-1)/2) * 2)+1

    half_win = (odd_win+1)/2-1

    tt = np.arange(0, number_pts)*resolution/sampl_freq
    ff = np.arange(0, win_size)*(sampl_freq/2)/(win_size-1)

    signal_padded = np.concatenate([np.zeros(odd_win-1), signal_in, np.zeros(odd_win-1)])

    wv = np.zeros((win_size, number_pts))

    r = np.zeros(win_size, dtype=complex)
    idx = np.arange(1, half_win + 1, dtype=int)

    for n in range(0,  int(number_pts/2)):

        t = 2 * n * resolution + odd_win -1
        r[0] = signal_padded[t] * np.conj(signal_padded[t])+1j*signal_padded[t+resolution] * \
            np.conj(signal_padded[t+resolution])
        v1 = signal_padded[t+idx]*np.conj(signal_padded[t-idx])
        v2 = signal_padded[t+resolution + idx]*np.conj(signal_padded[t+resolution-idx])
        r[idx] = v1+1j*v2
        r[win_size-idx] = np.conj(v1)+1j*np.conj(v2)

        r_fft = np.fft.fft(r, win_size)

        wv[:, 2*n] = np.real(r_fft)
        wv[:, 2*n+1] = np.imag(r_fft)

    return wv, ff, tt

File: 3_Signalanalyse/MATLAB/7_WVD/wvd.m
function [wv, ff, tt] = wvd(signalIn, resolution, winSize, smplFrequ)
% WVD creates a Wigner-Ville spectrogram 
%
% Input parameters:
% signalIn: real input time series
% resolution: resolution, number of samples between windows 
% (for full resolution:  1)
% winSize: window size, becomes length of frequency axis
% (a good default: length(signalIn)/2)
% smplFrequ: samples per second of signal
%
% Output parameters:
% wv: the W-V spectrum, each row represents a frequency, each column a 
% time instant
% ff: frequency vector (optional)
% tt: time vector (optional)
% 	
% Display using: 
%
% imagesc(tt,ff,wv);
% axis xy
% 	
% or:		
%
% surf(tt,ff,wv);
% shading interp
%
% Case Bradford, February 2005, http://case.caltech.edu/tfr/)
% Adapted from Rene Laterveer, 1999

if size(signalIn,1) ~= 1
    signalIn = transpose(signalIn);
end

% make even number of points, at given resolution
npts = floor(floor(length(signalIn) / resolution) / 2) * 2;

% make sure that we entered in an integer for the window
winSize = floor(winSize);

% round window length down to nearest odd integer
oddwin = (floor((winSize - 1) / 2) * 2) + 1;

% half point (for indexing reasons we need it later, we're
% filling two columns per loop, so we only index through half)
halfwin = (oddwin + 1) / 2 - 1;

% create tt and ff
tt = (0:npts - 1) * resolution / smplFrequ;
ff = (0:winSize - 1) * (smplFrequ / 2) / (winSize - 1);

% pad with zeros
signalIn = [zeros(1, oddwin - 1), signalIn, zeros(1, oddwin - 1)];

% initialize 
wv = zeros(winSize, npts);

r = zeros(1, winSize);
idx = 1:halfwin;

for n = 0 : npts/2 - 1
    
    t = 2 * n * resolution + oddwin;
    r(1) = signalIn(t) * conj(signalIn(t)) + 1i * ...
        signalIn(t + resolution) * conj(signalIn(t + resolution));
    v1 = signalIn(t + idx) .* conj(signalIn(t - idx));
    v2 = signalIn(t + resolution + idx) .* ...
        conj(signalIn(t + resolution - idx));
    r(idx + 1) = v1 + 1i * v2;
    r(winSize - idx + 1) = conj(v1) + 1i * conj(v2);
    
    rf = fft(r, winSize);
    
    wv(:, 2 * n + 1) = real(rf);
    wv(:, 2 * n + 2) = imag(rf);
    
end

end
